docs/design/2021-04-21-unify-log-library.md

```diff
@@ -0,0 +1,234 @@
+- Duplicate logs, e.g. `pd_stderr` will emit both text and json logs with duplicate content (but with a few subtle differences in timestamps).
+
+## Logging code for each component
+
+### [pingcap/log](https://github.com/pingcap/log)
+
+As a common logging library for PingCAP golang projects, it does the following things:
+
+- Provides the standard config schema.
+- Provides a factory method for creating log handlers.
+- Hard code the log format according to [unified-log-format RFC standard](https://github.com/tikv/rfcs/blob/master/text/2018-12-19-unified-log-format.md).
+- Encapsulates the logic of the rolling file.
+- Provides global log handler and related methods for package dimension.
+
+`pingcap/log` once had a strong limitation, that it couldn't customize the encoder for the text format. This problem had been [fixed by @9547](https://github.com/pingcap/log/pull/14).
+
+When PD and TiDB-operator were using `pingcap/log`, there was no custom encoder function yet, so they implemented one by themselves respectively, but accidentally wrote out a circular dependency. Currently, customize encoder [implemented by @9547](https://github.com/tikv/pd/pull/3480) and thus we can clear this tech debt.
+
+### TiDB
+
+Log library dependencies:
+
+![tidb-log-dependency](./imgs/tidb-log-dependency.png)
+
+For historical reasons, TiDB has two third-party logging libraries, `logrus` and `pingcap/log`. `pingcap/log` is a wrapper of `zap`.
+
+Logs of TiDB can be divided into two types, slow query logs and the other logs. 
+As mentioned above, these two types of logs are emitted through two different logging libraries, which results in separate configurations for the two types of logs and requires writing additional configuration conversion code.
+
+TiDB-specific logging logic is written inside `util/logutil/log.go`, e.g., logger initialization, logger configuration, and so on.
+
+Note this file, which is one of the main culprits of circular dependencies. The following briefly describes the key logic in `util/logutil/log.go`, two init methods and four log handlers.
+
+#### logrus
+
+The init method of `logrus` may initialize two `logrus` handlers.
+
+First, it is necessary to initialize the standard log handler (package level handler). `initLogger` first initializes the standard logger according to the configuration.
+
+```go
+func InitLogger(cfg *LogConfig) error {
+   log.SetLevel(stringToLogLevel(cfg.Level))
+   log.AddHook(&contextHook{})
+
+   if cfg.Format == "" {
+      cfg.Format = DefaultLogFormat
+   }
+   formatter := stringToLogFormatter(cfg.Format, cfg.DisableTimestamp)
+   log.SetFormatter(formatter)
+
+   if len(cfg.File.Filename) != 0 {
+      if err := initFileLog(&cfg.File, nil); err != nil {
+         return errors.Trace(err)
+      }
+   }
+
+// The rest is omitted.
+```
```

Comment:
Can we use other methods (maybe a link) to display this code? They make the document longer.